app/src/main/java/org/kiwix/kiwixmobile/local_file_transfer/WifiDirectManager.kt

```diff
@@ -0,0 +1,395 @@
+
+  fun getFilesForTransfer() = filesForTransfer
+
+  fun setFilesForTransfer(fileItems: ArrayList<FileItem>) {
+    filesForTransfer = fileItems
+  }
+
+  val zimStorageRootPath
+    get() = sharedPreferenceUtil.prefStorage + "/Kiwix/"
+
+  fun getFileReceiverDeviceAddress() = fileReceiverDeviceAddress
+
+  fun setClientAddress(clientAddress: InetAddress) {
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress
+    startFileTransfer()
+  }
+
+  private fun startFileTransfer() {
+    if (isGroupFormed) {
+      if (isFileSender) {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer")
+        fileReceiverDeviceAddress =
+          if (isGroupOwner) selectedPeerDeviceInetAddress else groupOwnerAddress
+        displayToast(R.string.preparing_files, Toast.LENGTH_LONG)
+        senderDeviceAsyncTask = SenderDeviceAsyncTask(this, activity).also {
+          it.execute()
+        }
+      } else {
+        callbacks.onFilesForTransferAvailable(filesForTransfer)
+        receiverDeviceAsyncTask = ReceiverDeviceAsyncTask(this).also {
+          it.execute()
+        }
+      }
+    }
+  }
+
+  fun changeStatus(itemIndex: Int, @FileStatus status: Int) {
+    filesForTransfer[itemIndex].fileStatus = status
+    callbacks.onFileStatusChanged(itemIndex)
+    if (status == FileStatus.ERROR) {
+      displayToast(
+        R.string.error_transferring, filesForTransfer[itemIndex].fileName,
+        Toast.LENGTH_SHORT
+      )
+    }
+  }
+
+  private fun cancelAsyncTasks(vararg tasks: AsyncTask<*, *, *>?) =
+    tasks.all {
+      it?.cancel(true) ?: false
+    }
+
+  fun stopWifiDirectManager() {
+    cancelAsyncTasks(peerGroupHandshakeAsyncTask, senderDeviceAsyncTask, receiverDeviceAsyncTask)
+    if (isFileSender) {
+      closeChannel()
+    } else {
+      disconnect()
+    }
+    unregisterWifiDirectBroadcastReceiver()
+  }
+
+  private fun disconnect() {
+    manager.removeGroup(channel, object : ActionListener {
+      override fun onFailure(reasonCode: Int) {
+        Log.d(TAG, "Disconnect failed. Reason: $reasonCode")
+        closeChannel()
+      }
+
+      override fun onSuccess() {
+        Log.d(TAG, "Disconnect successful")
+        closeChannel()
+      }
+    })
+  }
+
+  private fun closeChannel() {
+    if (VERSION.SDK_INT >= VERSION_CODES.O_MR1) {
+      channel.close()
+    }
+  }
+
+  fun getErrorMessage(reason: Int): String {
+    return when (reason) {
+      WifiP2pManager.ERROR -> "Internal error"
+      WifiP2pManager.BUSY -> "Framework busy, unable to service request"
+      WifiP2pManager.P2P_UNSUPPORTED -> "P2P unsupported on this device"
+      else -> "Unknown error code - $reason"
+    }
+  }
+
+  fun displayToast(
+    stringResourceId: Int,
+    templateValue: String,
+    duration: Int
+  ) = activity.toast(activity.getString(stringResourceId, templateValue), duration)
+
+  fun displayToast(stringResourceId: Int, duration: Int) =
```

Comment:
same for this

Reply:
This function is also used by ReceiveDeviceAsyncTask and activity is private so it can't be used by ReceiveDeviceAsyncTask.

Reply:
ah, annoying. We had planned to get rid of these asynctasks and replace them with Rx but it just never happened